CORE-7147: Validate transaction metadata JSON

omfR3 opened the pull request:
We use `TransactionMetaData` to represent ledger metadata for transactions, including information like cipher digest settings and CPI/CPK package details. When encoding a transaction, we serialise this information to JSON and its contents influence the transaction signature. However, we store the data within `TransactionMetaData` as a very loosely-typed `Map<String, Any?>`, which means Jackson can't do much validation for us.

This PR adds a custom JSON schema for this metadata and uses it to validate transactions during deserialisation.

## TODO

- [x] Remove duplication in `TransactionMetaData::entries` ([CORE-7145](https://r3-cev.atlassian.net/browse/CORE-7145))
- [x] Generate JSON schema
- [x] Validate JSON against schema when deserialising
- [x] Canonicalise JSON when encoding metadata
- [x] Check that serialised metadata is canonicalised when decoding
- [x] Figure out what to do about broken OSGi
- [x] Resolve approximately 1 jillion merge conflicts
- [x] Make single schema that can validate both UTXO and consensual ledger transactions
- [ ] Possibly ~do a big refactoring to introduce a factory and reduce some blind-passthrough dependency coupling~ (no, @vlajos is working on this in a separate PR)
- [x] Add two missing bits of metadata (`schemaVersion` and `transactionSubtype`)
- [x] Fix incorrect representation of `transactionSubtype` (missing optional/conditional fields should be `null` or not present when serialised, not the string "`null`")
- [x] Refactor out duplicate metadata creation in persistence tests
- [x] Rename digest settings keys from `SNAKE_CONST_CASE` to `camelCase` for consistency

omfR3 committed:
CORE-7147: Add JSON schema for transaction metadata

* Add JSON schema based on current serialised form of transaction
  metadata
* Note, this is in ledger-consensual-impl for now but should probably be
  moved to ledger-common at some point
* Also remove unused "entries" property from TransactionMetaData

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
CORE-7147: (ledger-consensual-impl) Reorder TX metadata schema required fields to be more pleasing + fix merge error

omfR3 committed:
CORE-7147: (ledger) Add a couple of descriptions to the transaction metadata JSON schema + other text improvements

omfR3 committed:
CORE-7147: (ledger-common) Move TX metadata JSON schema to ledger-common-impl

vlajos commented on libs/ledger/ledger-common-impl/src/main/resources/schema/transaction-metadata-v1.json:
```diff
@@ -0,0 +1,241 @@
+				},
+				"ROOT_MERKLE_TREE_DIGEST_OPTIONS_LEAF_PREFIX_B64": {
+					"$id": "#root/digestSettings/ROOT_MERKLE_TREE_DIGEST_OPTIONS_LEAF_PREFIX_B64", 
+					"title": "Root Merkle tree digest options leaf prefix b64", 
```
```suggestion
					"title": "Root Merkle tree digest options leaf prefix base64 encoded", 
```

vlajos commented on libs/ledger/ledger-common-impl/src/main/resources/schema/transaction-metadata-v1.json:
```diff
@@ -0,0 +1,241 @@
+				},
+				"ROOT_MERKLE_TREE_DIGEST_OPTIONS_NODE_PREFIX_B64": {
+					"$id": "#root/digestSettings/ROOT_MERKLE_TREE_DIGEST_OPTIONS_NODE_PREFIX_B64", 
+					"title": "Root Merkle tree digest options node prefix b64", 
```
```suggestion
					"title": "Root Merkle tree digest options node prefix base64 encoded", 
```

vlajos commented on libs/ledger/ledger-common-impl/src/main/resources/schema/transaction-metadata-v1.json:
```diff
@@ -0,0 +1,241 @@
+			}
+		}
+,
+		"cpkMetadata": {
```
Is not it possible to define somehow only once the `CordaPackageSummary` ones?

omfR3 replied:
Yes, good idea. I've extracted the common structure to the `definitions` section at the bottom.

vlajos replied:
Are the `$id`s OK in that section? They contain `cpi`. Also the examples do not seem to be very realistic.

omfR3 replied:
Oh you're right, I should probably remove the `$id` fields (they're autogenerated).

omfR3 committed:
Update libs/ledger/ledger-common-impl/src/main/resources/schema/transaction-metadata-v1.json

Co-authored-by: Lajos Veres <lajos.veres@r3.com>

omfR3 committed:
Update libs/ledger/ledger-common-impl/src/main/resources/schema/transaction-metadata-v1.json

Co-authored-by: Lajos Veres <lajos.veres@r3.com>

omfR3 committed:
CORE-7147: (ledger) TX metadata JSON schema: Refactor out common package definition

omfR3 committed:
CORE-7147: (ledger) TX metadata JSON schema: remove $id fields and improve example data

omfR3 committed:
CORE-7147: (ledger) Add JsonValidator for schema validation of transaction metadata

* Add JsonValidator interface + impl
* Move mockTransactionMetaData() from ConsensualTransactionMocks to
  WireTransactionExample and add required Gradle dependencies

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
CORE-7147: (ledger) Canonicalise TX metadata JSON + check during validation

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

lankydan commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/WireTransaction.kt:
```diff
@@ -137,6 +138,12 @@ class WireTransaction(
     }
+
+    private fun parseMetadata(json: String): TransactionMetaData {
+        JsonValidatorImpl().validate(json, schemaPath = "/schema/transaction-metadata-v1.json")
```
I think we'd need this to be injected to resolve your osgi issue. Constructing an implementation here seems a bit off anyway.

omfR3 replied:
I actually started out with the intention of making it injectable (that's why there's a `JsonValidator` interface), but we wanted to keep things simple first unless it became really necessary to complicate things. Seems that it is really necessary, though.

What bugs me about it is that we're passing down lots of dependencies via constructor injection from several levels above `WireTransaction`. This means those classes would all need to be updated to receive and pass down a `jsonValidator: JsonValidator` parameter which just feels like a bunch of unnecessary coupling.

omfR3 committed:
CORE-7147: (ledger) Injectify JsonValidator. Manually include json-canonicalisation JAR in ledger-common-data bundle.

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

vlajos commented on components/ledger/ledger-consensual-flow/src/main/kotlin/net/corda/ledger/consensual/flow/impl/transaction/ConsensualTransactionBuilderImpl.kt:
```diff
@@ -103,9 +106,8 @@ class ConsensualTransactionBuilderImpl(
-                            .toByteArray(Charsets.UTF_8)
-                    ) // TODO(update with CORE-6890)
+                        serializeMetadata().toByteArray(Charsets.UTF_8)
+                    ) // TODO(update with CORE-5940)
```
I guess this todo is not relevant anymore

vlajos commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/WireTransaction.kt:
```diff
@@ -36,7 +39,7 @@ class WireTransaction(
 
         val metadataBytes = componentGroupLists[ALL_LEDGER_METADATA_COMPONENT_GROUP_ID].first()
         // TODO(update with CORE-6890)
```
todo looks done.

vlajos commented on testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/WireTransactionExample.kt:
```diff
@@ -1,33 +1,66 @@
+
     val componentGroupLists = listOf(
-        listOf(jsonMarshallingService.format(transactionMetaData).toByteArray(Charsets.UTF_8)), // TODO(update with CORE-6890)
+        listOf(canonicalJson.toByteArray(Charsets.UTF_8)), // TODO(update with CORE-6890)
```
Todo is done now.

vlajos commented on libs/ledger/ledger-common-data/build.gradle:
```diff
@@ -36,3 +46,23 @@ dependencies {
+
+artifacts {
+    bundle jar
+}
```
If it is not too complicated probably it would worth extracting this into a dedicated module.

omfR3 committed:
CORE-7147: Remove old TODOs

omfR3 committed:
CORE-7147: Move JsonValidator into new project :libs:serialization:json-validator

omfR3 commented:
I've extracted `JsonValidator` into its own project in `:libs:serialization:json-validator` and it seems happy there, except that at least one integration test fails when trying to read the schema file with `getResourceAsStream`:

```
TEST net.corda.ledger.common.flow.impl.transaction.serializer.kryo.tests.WireTransactionKryoSerializationTest#correct serialization of a wire Transaction() <<< ERROR: Couldn't load JSON schema from /schema/tran
saction-metadata-v1.json
net.corda.v5.base.exceptions.CordaRuntimeException: Couldn't load JSON schema from /schema/transaction-metadata-v1.json
        at net.corda.common.json.validation.impl.JsonValidatorImpl.validateSchema(JsonValidatorImpl.kt:33)
        at net.corda.common.json.validation.impl.JsonValidatorImpl.validate(JsonValidatorImpl.kt:18)
        at net.corda.ledger.common.data.transaction.WireTransaction.parseMetadata(WireTransaction.kt:144)
        at net.corda.ledger.common.data.transaction.WireTransaction.<init>(WireTransaction.kt:41)
        at net.corda.ledger.common.testkit.WireTransactionExampleKt.getWireTransaction(WireTransactionExample.kt:28)
        at net.corda.ledger.common.flow.impl.transaction.serializer.kryo.tests.WireTransactionKryoSerializationTest.correct serialization of a wire Transaction(WireTransactionKryoSerializationTest.kt:86)
[...]
> Task :components:ledger:ledger-common-flow:testOSGi FAILED
```

Right now that schema lives at `libs/ledger/ledger-common-data/src/main/resources/schema/transaction-metadata-v1.json`, so I assumed it would be available to anything that depends on `:libs:ledger:ledger-common-data`, but no. Also, maybe it should live in some other project that just holds JSON schemas. But the schema is definitely inside that JAR:

```
$ unzip -l libs/ledger/ledger-common-data/build/libs/corda-ledger-common-data-5.0.0.0-SNAPSHOT.jar

Archive:  libs/ledger/ledger-common-data/build/libs/corda-ledger-common-data-5.0.0.0-SNAPSHOT.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
     ....
        0  2022-10-24 16:46   schema/
     4366  2022-10-24 12:13   schema/transaction-metadata-v1.json
```

@lankydan @vlajos do you have any tips on this sort of thing? I'm not sure how to debug this sort of classpath / JAR resources trickery :dizzy_face: 

[edit]
Same thing happens in `:libs:ledger:ledger-common-data:integrationTest`, even though I've tried adding `schema/transaction-metadata-v1.json` to both `src/test/resources` and `src/integrationTest/resources`. Maybe I'm missing something in `test.bndrun` or `build.gradle`?

lankydan commented on libs/ledger/ledger-common-data/src/main/resources/schema/transaction-metadata-v1.json:
```diff
@@ -0,0 +1,178 @@
+{
+	"$schema": "http://json-schema.org/draft-07/schema#", 
```
Can we make this file less indented 😄 ? Seems like 8 spaces instead of 4.

omfR3 replied:
Yes -- there also seemed to be tabs in it from the schema generator tool I used. Will push a reindented version later.

lankydan commented on libs/ledger/ledger-common-data/src/main/resources/schema/transaction-metadata-v1.json:
```diff
@@ -0,0 +1,178 @@
+				"COMPONENT_MERKLE_TREE_ENTROPY_ALGORITHM_NAME"
+			],
+			"properties": {
+				"ROOT_MERKLE_TREE_DIGEST_PROVIDER_NAME": {
```
I know this is how they show in the code atm, but should things like `ROOT_MERKLE_TREE_DIGEST_PROVIDER_NAME` be in capitals when everything else isn't?

omfR3 replied:
I'm not sure -- I had assumed they were written that way (in `WireTransactionDigestSettings.kt`) as an enum or something, or that the formatting had some other significance. Maybe @vlajos has a reasoning for it to be one way or the other?

vlajos replied:
I think I refactored `WireTransactionDigestSettings.kt` forth and back a few times and just checked if the keys were consistent with the constants' names.
Anyway I agree that the schema and the json would look nicer with consistent lowercase.

vlajos replied:
@omfR3  This topic looks still pending.

omfR3 replied:
Renamed these (hopefully without breaking everything).

lankydan commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/WireTransaction.kt:
```diff
@@ -137,6 +139,12 @@ class WireTransaction(
         merkleTreeProvider.createTree(componentGroupRoots, getRootMerkleTreeDigestProvider())
     }
+
+    private fun parseMetadata(json: String): TransactionMetaData {
```
I'm going to suggest a large refactoring here.

Can we refactor `ConsensualTransactionBuilderImpl` to include either a factory to create a `ConsensualSignedTransaction` or a `WireTransaction`? This factory should live in the `data` module since there isn't somewhere else to put it and I'm happy with that.

We should do this since `ConsensualTransactionBuilderImpl` passes a crap load of dependencies down into `ConsensualSignedTransaction` and `WireTransaction`. By doing this refactoring the builder impl itself becomes pretty small.

Personally, I think the best choice is making both a `ConsensualSignedTransactionFactory` and a `WireTransactionFactory`, since there is code in `WireTransaction` that should be moved out but we also need to ensure that it gets run when deserialized by our custom serializer.

Anyway, the main reason I wanted to do this, was so that we could move the `parseMetadata` method out of here and into the `WireTransactionFactory`. This means that both the `JsonValidator` and `JsonMarshallingService` can be removed from this class, thus simplifying the `WireTransaction` class itself.

lankydan replied:
Doing this might also mean that other modules do not need `JsonValidator` itself because the validation will be purely internal to the `WireTransactionFactory` (unless I've missed something).

omfR3 replied:
I do hate the way we're passing down all of those dependencies at like 3 levels of hierarchy. I'm a bit iffy about adding a big refactor to this PR, but I like the idea of removing some of the DI clutter (and unwanted coupling).

blsemo replied:
@lankydan Just for me to understand:
You popose that the checks/metadata deserialization should happen in a factory class that gets injected into the TX builder, so that only that factory will have the dependency on the JSON marshalling and validation services.
The custom serilalizers will have to get hold of that factory as well so they can create the objects when deseriallizing?

lankydan commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/WireTransaction.kt:
```diff
@@ -137,6 +139,12 @@ class WireTransaction(
     }
+
+    private fun parseMetadata(json: String): TransactionMetaData {
+        jsonValidator.validate(json, schemaPath = "/schema/transaction-metadata-v1.json")
```
Probably worth making the json path a constant.

vlajos replied:
@omfR3 This one looks pending as well

lankydan commented:
> I've extracted `JsonValidator` into its own project in `:libs:serialization:json-validator` and it seems happy there, except that at least one integration test fails when trying to read the schema file with `getResourceAsStream`:
> 
> ```
> TEST net.corda.ledger.common.flow.impl.transaction.serializer.kryo.tests.WireTransactionKryoSerializationTest#correct serialization of a wire Transaction() <<< ERROR: Couldn't load JSON schema from /schema/tran
> saction-metadata-v1.json
> net.corda.v5.base.exceptions.CordaRuntimeException: Couldn't load JSON schema from /schema/transaction-metadata-v1.json
>         at net.corda.common.json.validation.impl.JsonValidatorImpl.validateSchema(JsonValidatorImpl.kt:33)
>         at net.corda.common.json.validation.impl.JsonValidatorImpl.validate(JsonValidatorImpl.kt:18)
>         at net.corda.ledger.common.data.transaction.WireTransaction.parseMetadata(WireTransaction.kt:144)
>         at net.corda.ledger.common.data.transaction.WireTransaction.<init>(WireTransaction.kt:41)
>         at net.corda.ledger.common.testkit.WireTransactionExampleKt.getWireTransaction(WireTransactionExample.kt:28)
>         at net.corda.ledger.common.flow.impl.transaction.serializer.kryo.tests.WireTransactionKryoSerializationTest.correct serialization of a wire Transaction(WireTransactionKryoSerializationTest.kt:86)
> [...]
> > Task :components:ledger:ledger-common-flow:testOSGi FAILED
> ```
> 
> Right now that schema lives at `libs/ledger/ledger-common-data/src/main/resources/schema/transaction-metadata-v1.json`, so I assumed it would be available to anything that depends on `:libs:ledger:ledger-common-data`, but no. Also, maybe it should live in some other project that just holds JSON schemas. But the schema is definitely inside that JAR:
> 
> ```
> $ unzip -l libs/ledger/ledger-common-data/build/libs/corda-ledger-common-data-5.0.0.0-SNAPSHOT.jar
> 
> Archive:  libs/ledger/ledger-common-data/build/libs/corda-ledger-common-data-5.0.0.0-SNAPSHOT.jar
>   Length      Date    Time    Name
> ---------  ---------- -----   ----
>      ....
>         0  2022-10-24 16:46   schema/
>      4366  2022-10-24 12:13   schema/transaction-metadata-v1.json
> ```
> 
> @lankydan @vlajos do you have any tips on this sort of thing? I'm not sure how to debug this sort of classpath / JAR resources trickery 😵
> 
> [edit] Same thing happens in `:libs:ledger:ledger-common-data:integrationTest`, even though I've tried adding `schema/transaction-metadata-v1.json` to both `src/test/resources` and `src/integrationTest/resources`. Maybe I'm missing something in `test.bndrun` or `build.gradle`?

I assume its because the integration tests are using the OSGi/bundled jars so they wouldn't include anything from the integrationTest resources directories themselves.

But in regards to why it can't find it generally between modules I'm not super sure atm

lankydan commented:
I think the factory refactor can solve this, because we can also decide to bypass the json validation in the testkit code. A factory can perform the validation upon creation but we could do something else for tests.

lankydan commented on libs/serialization/json-validator/src/main/kotlin/net/corda/common/json/validation/impl/JsonValidatorImpl.kt:
```diff
@@ -0,0 +1,38 @@
+package net.corda.common.json.validation.impl
+
+import net.corda.v5.base.exceptions.CordaRuntimeException
```
imports need reodering

lankydan commented on libs/serialization/json-validator/src/main/kotlin/net/corda/common/json/validation/impl/JsonValidatorImpl.kt:
```diff
@@ -0,0 +1,38 @@
+        val schema = JsonSchemaFactory
+            .getInstance(SpecVersion.VersionFlag.V7)
+            .getSchema(
+                javaClass.getResourceAsStream(schemaPath)
```
@driessamyn @chrisr3 is `getResourceAsSteam` something that we are allowed to do in OSGi and bundle land or should we be using something else?

chrisr3 replied:
> is `getResourceAsSteam` something that we are allowed to do in OSGi and bundle land or should we be using something else?

It's unlikely to find what you're looking for unless it's a resource inside the same bundle.

omfR3 replied:
Thanks @chrisr3, that's what I was beginning to suspect. I've changed it to load the resource at the call site and pass the `InputStream` instead of a path. Feels a bit smelly but I don't see a significantly less bad alternative.

omfR3 replied:
I guess we could put all internal JSON schemas inside the `json-validator` project where it can access them, and export some constants, so you could call it like:
```kotlin
  jsonValidator.validate(json, JsonSchemas.CONSENSUAL_LEDGER_METADATA)
```
...with the constants defined in one place like this:
```kotlin
class JsonSchemas {
  companion object {
    const val CONSENSUAL_LEDGER_METADATA = "/schema/consensual-transaction-metadata-v1.json"
    const val SOME_OTHER_SCHEMA = "/schema/wabbits-v1.json"
  }
}
```

I guess it's a question of whether we want to centralise that information in the one file, or deal with loading resources at the call sites (of which there's only 1 currently).

What think ye, @driessamyn @chrisr3 @lankydan?

omfR3 replied:
Yeah I like that more actually. Ok, I'm doin' it.

(Update: I've backed this out and put the schema in `ledger-common-data` for now since we're using it to validate both UTXO and consensual transactions)

omfR3 committed:
CORE-7147: (ledger) Reformat ugly JSON schema + make JsonValidator take an InputStream rather than path to schema, since it won't be able to directly reference a resource within the calling class's JAR.

omfR3 committed:
CORE-7147: Undo changes to JsonValidator + add JsonSchemas

* JsonValidator: take schema path, not inputstream.
* Move consensual transaction metadata JSON schema to json-validator project
* Add JsonSchemas with static constants, rather than expecting callers
  to hardcode paths

omfR3 committed:
CORE-7147: Fix a broken integration test

omfR3 committed:
CORE-7147: Reject unknown properties in consensual ledger metadata JSON schema

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
CORE-7147: (json-validator) Exclude transitive dependency on vulnerable Jackson version

omfR3 committed:
CORE-7147: (ledger) Add JSON validation to UTXO ledger transactions

* move ConsensualTransactionMetaDataExample to
  TransactionMetaDataExample in ledger-common-testkit which can be used
  by both ledgers
* validate/canonicalise metadata in UTXO transactions too
* rm unused file

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
CORE-7147: (ledger) Validate consensual/UTXO metadata with 1 JSON schema

* Switch metadata schema to draft/2019-09 version
* Add conditional clauses to schema so it can validate both consensual
  and UTXO metadata (currently, UTXO just has one extra required field)
* Move schema back from json-validator -> ledger-common-data and get rid
  of JsonSchemas.kt
* Make schema ledgerModel an enum, currently allowing the canonical
  classnames for Consensual/UTXOTransactionImpl classes.

omfR3 committed:
CORE-7147: (ledger) Fix typo and inject a missing service

omfR3 committed:
CORE-7147: (ledger metadata) Add enum for UTXO transaction subtypes

* Move TRANSACTION_META_DATA_UTXO_LEDGER_VERSION const into new
  UtxoTransactionMetaData class as static const, rename to
  LEDGER_VERSION.
* Add TransactionSubtype enum with two subtypes.
* Add getter method "getTransactionSubtype" to TransactionMetaData so
  Jackson knows how to serialise it to JSON.

omfR3 committed:
CORE-7147: (json-validator) Use correct schema draft version

omfR3 committed:
CORE-7147: (ledger inttest) Add missing properties to sample metadata + canonicalise JSON

vlajos commented on components/ledger/ledger-consensual-persistence-impl/src/integrationTest/kotlin/net/corda/ledger/consensual/persistence/impl/processor/tests/ConsensualLedgerRepositoryTest.kt:
```diff
@@ -370,14 +376,17 @@ class ConsensualLedgerRepositoryTest {
+            TransactionMetaData.CPI_METADATA_KEY to cpiPackageSummaryExample,
+            TransactionMetaData.CPK_METADATA_KEY to cpks
+        ))
+
```
Could we use here the metadata example file?

omfR3 replied:
I would prefer that, but figured there was some special behaviour happening here with the CPK information. @mbrkic-r3 is this the case, or can we use the sample metadata in the testkits?

omfR3 replied:
I've resolved one of the duplicate metadata bits in the repository tests, but not this one since the custom CPK setup is important for the test here.

vlajos commented on libs/ledger/ledger-common-data/src/main/resources/schema/transaction-metadata.json:
```diff
@@ -0,0 +1,202 @@
+      "type": "string",
+      "default": "",
+      "examples": [
+        "0.001"
```
Can you update this version number as well?

omfR3 replied:
I'll change all of them to match this (since that's what you used in the UTXO sample metadata).

vlajos replied:
I am fine with any of the formats, just we should be consistent.

omfR3 replied:
Yep, I agree.

vlajos commented on testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/WireTransactionExample.kt:
```diff
@@ -16,20 +19,53 @@ fun getWireTransactionExample(
+                    "0303030303030303030303030303030303030303030303030303030303030303")
+            )
+        )
+    )
```
Is not this the same as testing/ledger/ledger-common-testkit/src/main/kotlin/net/corda/ledger/common/testkit/TransactionMetaDataExample.kt ?

omfR3 replied:
It's not used anymore anyway so I'll remove it now.

vlajos commented:
What did happen with testing/ledger/ledger-utxo-testkit/src/main/kotlin/net/corda/ledger/utxo/testkit/UtxoSignedTransactionExample.kt?

vlajos commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/impl/transaction/UtxoTransactionMetaData.kt:
```diff
@@ -1,3 +1,12 @@
+        GENERAL
+    }
+    companion object {
+        const val LEDGER_VERSION = "0.001"
```
Could you change the consensual one's version accordingly to keep them similar?

omfR3 replied:
@mnesbit suggested that ledger version should be an integer instead, so I'll change them all.

omfR3 commented:
> What did happen with testing/ledger/ledger-utxo-testkit/src/main/kotlin/net/corda/ledger/utxo/testkit/UtxoSignedTransactionExample.kt?

Ah yes I meant to say something about that -- it just exported one function that didn't seem to be used anywhere, and would need to add a dependency on `json-validator` to compile, so I thought better to delete it for now. If you have a need for it, I can bring it back and add that dependency.

vlajos commented:
> > What did happen with testing/ledger/ledger-utxo-testkit/src/main/kotlin/net/corda/ledger/utxo/testkit/UtxoSignedTransactionExample.kt?
  > 
  > Ah yes I meant to say something about that -- it just exported one function that didn't seem to be used anywhere, and would need to add a dependency on `json-validator` to compile, so I thought better to delete it for now. If you have a need for it, I can bring it back and add that dependency.

They will needed for the serializers. Could you please add it back?

omfR3 committed:
CORE-7147: (ledger) Improvements from PR comments

* Restore deleted UtxoSignedTransactionExample.kt (will be needed soon
  for serialisers)
* Make ledgerVersion in various TX metadata examples more consistent
* Delete no-longer used mock data function

omfR3 committed:
CORE-7147: (ledger) Change tx metadata ledgerVersion from String -> Int

mnesbit commented:
I am fine with the framework here. I do feel we need to look a bit at the metadata and thence the JSON schema to make sure it is doing what we need.

omfR3 commented:
Good point @mnesbit -- here's the current JSON output for the metadata in one of the consensual ledger unit tests and one of the UTXO tests. Now that I look at it, the CPI/CPK version is a string here. We said it should be an int, but that doesn't match `CpkIdentifier` which takes a string. Not sure what to do about that discrepancy.

## Consensual transaction metadata

```json
{
  "cpiMetadata": {
    "fileChecksum": "416E6F746865722D46616B652D76616C7565",
    "name": "CPI name",
    "signerSummaryHash": "46616B652D76616C7565",
    "version": "1"
  },
  "cpkMetadata": [
    {
      "fileChecksum": "0101010101010101010101010101010101010101010101010101010101010101",
      "name": "MockCpk",
      "signerSummaryHash": "",
      "version": "1"
    },
    {
      "fileChecksum": "0303030303030303030303030303030303030303030303030303030303030303",
      "name": "MockCpk",
      "signerSummaryHash": "",
      "version": "3"
    }
  ],
  "digestSettings": {
    "COMPONENT_MERKLE_TREE_DIGEST_ALGORITHM_NAME": "SHA-256D",
    "COMPONENT_MERKLE_TREE_DIGEST_PROVIDER_NAME": "NonceHashDigestProvider",
    "COMPONENT_MERKLE_TREE_ENTROPY_ALGORITHM_NAME": "SHA-256D",
    "ROOT_MERKLE_TREE_DIGEST_ALGORITHM_NAME": "SHA-256D",
    "ROOT_MERKLE_TREE_DIGEST_OPTIONS_LEAF_PREFIX_B64": "bGVhZg==",
    "ROOT_MERKLE_TREE_DIGEST_OPTIONS_NODE_PREFIX_B64": "bm9kZQ==",
    "ROOT_MERKLE_TREE_DIGEST_PROVIDER_NAME": "TweakableHashDigestProvider"
  },
  "ledgerModel": "net.corda.ledger.consensual.data.transaction.ConsensualLedgerTransactionImpl",
  "ledgerVersion": 1,
  "schemaVersion": 1,
  "transactionSubtype": null
}
```

## UTXO transaction metadata

```json
{
  "cpiMetadata": {
    "fileChecksum": "416E6F746865722D46616B652D76616C7565",
    "name": "CPI name",
    "signerSummaryHash": "46616B652D76616C7565",
    "version": "1"
  },
  "cpkMetadata": [
    {
      "fileChecksum": "0101010101010101010101010101010101010101010101010101010101010101",
      "name": "MockCpk",
      "signerSummaryHash": "",
      "version": "1"
    },
    {
      "fileChecksum": "0303030303030303030303030303030303030303030303030303030303030303",
      "name": "MockCpk",
      "signerSummaryHash": "",
      "version": "3"
    }
  ],
  "digestSettings": {
    "COMPONENT_MERKLE_TREE_DIGEST_ALGORITHM_NAME": "SHA-256D",
    "COMPONENT_MERKLE_TREE_DIGEST_PROVIDER_NAME": "NonceHashDigestProvider",
    "COMPONENT_MERKLE_TREE_ENTROPY_ALGORITHM_NAME": "SHA-256D",
    "ROOT_MERKLE_TREE_DIGEST_ALGORITHM_NAME": "SHA-256D",
    "ROOT_MERKLE_TREE_DIGEST_OPTIONS_LEAF_PREFIX_B64": "bGVhZg==",
    "ROOT_MERKLE_TREE_DIGEST_OPTIONS_NODE_PREFIX_B64": "bm9kZQ==",
    "ROOT_MERKLE_TREE_DIGEST_PROVIDER_NAME": "TweakableHashDigestProvider"
  },
  "ledgerModel": "net.corda.ledger.utxo.data.transaction.UtxoLedgerTransactionImpl",
  "ledgerVersion": 1,
  "schemaVersion": 1,
  "transactionSubtype": "GENERAL"
}
```

omfR3 committed:
CORE-7147: (ledger) Add schemaVersion to TX metadata

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
CORE-7147: (ledger) Fix/refactor TX metadata tests

* Remove some duplicate code for creating example metadata
* Adding missing schema version to test metadata
* Make conditional metadata property "transactionSubtype" nullable
  rather than producing the string "null" when it's missing

omfR3 committed:
CORE-7147: (ledger) Rename digest settings constants from SNAKE_CONST_CASE to camelCase for consistency

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
CORE-7147: Add missing @Reference in ConsensualLedgerRepository

omfR3 commented:
I had a discussion about this with @chrisr3 who pointed out that we can't abide by the creation of an `ObjectMapper` inside `JsonValidatorImpl`. In theory we could work around that by exposing some extra functionality in `JsonMarshallingService` (or move the validation functionality into that service).
However, we can't really do anything about the schema validator library internally creating more instances of `ObjectMapper` when parsing the schema file.

Accordingly I'm going to close this PR for now because there's no way we can merge it.

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 commented:
After some more discussion, I don't think the `ObjectMapper` issue is a problem here, since it's not used to deserialise JSON into arbitrary classes. Instead, it just creates `JsonNode` instances for working directly with the object tree.

omfR3 committed:
CORE-7147: Detekt fix + workaround

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
CORE-7147: (ledger persistence) Reduce code duplication with example TX metadata

* Change transactionMetadataExample into a function with optional args.
* Use the above function in ConsensualLedgerRepositoryTest since only
  the CPK metadata is different.

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-7147/validate-transaction-metadata-json

omfR3 committed:
CORE-7147: Detekt

vlajos commented on components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/factory/TransactionMetadataFactoryImpl.kt:
```diff
@@ -24,12 +24,13 @@ class TransactionMetadataFactoryImpl @Activate constructor(
             TransactionMetadata.CPI_METADATA_KEY to getCpiSummary(),
-            TransactionMetadata.CPK_METADATA_KEY to getCpkSummaries()
+            TransactionMetadata.CPK_METADATA_KEY to getCpkSummaries(),
+            TransactionMetadata.SCHEMA_VERSION_KEY to 1
```
I think this `1` could be a constant.

omfR3 committed:
CORE-7147: (ledger metadata) Move metadata schema version to a constant

lankydan commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/TransactionMetadata.kt:
```diff
@@ -32,7 +33,18 @@ class TransactionMetadata(private val properties: LinkedHashMap<String, Any>) {
+        val version = this[LEDGER_VERSION_KEY].toString()
+
+        try {
+            return Integer.parseInt(version)
```
I think you can do `toInt()` afaik

lankydan commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/TransactionMetadata.kt:
```diff
@@ -32,7 +33,18 @@ class TransactionMetadata(private val properties: LinkedHashMap<String, Any>) {
+            return Integer.parseInt(version)
+        } catch (e: NumberFormatException) {
+            throw CordaRuntimeException(
+            "Transaction metadata representation error: ledger version should be an integer but could not be parsed: $version")
```
Formatting is off here.

lankydan commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/factory/WireTransactionFactoryImpl.kt:
```diff
@@ -81,6 +84,9 @@ class WireTransactionFactoryImpl @Activate constructor(
         return jsonMarshallingService.parse(metadataBytes.decodeToString(), TransactionMetadata::class.java)
     }
 
+    private fun getSchema(path: String) =
```
This means we are loading the resource each time we create a `WireTransaction`.

Should it be a lazy property of the factory? If it's being loaded as a resource, then it can't change since its part of the jar that we're running. This means we initialise it once instead of many times.

omfR3 replied:
That's right. I wasn't sure how best to handle it, but putting it in the factory might be a good idea. Although it feels a bit "leaky".

omfR3 replied:
@lankydan how about this change? The factory refactoring actually made it really easy since its lifetime is naturally decoupled from each instance of `WireTransaction`.

omfR3 committed:
CORE-7147: Improvements from PR feedback

* Cleanup in TransactionMetadata
* JsonValidator: separate schema parsing from validation + add
  WrappedJsonSchema class to avoid leaking library dependency
* WireTransactionFactoryImpl: load schema once, lazily, to avoid
  repeating that work every time a WireTransaction is created.

vlajos commented on libs/serialization/json-validator/src/main/kotlin/net/corda/common/json/validation/WrappedJsonSchema.kt:
```diff
@@ -0,0 +1,5 @@
+
+import com.networknt.schema.JsonSchema
+
+data class WrappedJsonSchema(val schema: JsonSchema)
```
Why is this wrapper needed?

omfR3 replied:
Without the wrapper, `WireTransactionFactoryImpl` (and any other caller) will need to add a Gradle dependency on the `com.networknt:json-schema-validator` library, which is a bit ugly because it needs an exclusion clause for an outdated transitive dependency. I guess we could make it an `api` dependency in the `json-validator` project and that might make it directly visible, but I'm not sure we should leak `com.networknt:json-schema-validator` out just for this.

(edit) Note that the calling party doesn't actually access anything in the `WrappedJsonSchema` and therefore doesn't need to see the `JsonSchema` API -- it just holds onto it, then passes it to `jsonValidator.validate(...)` when needed.

omfR3 replied:
(I described the rationale in the commit message, BTW)

vlajos replied:
I see. Thanks.

lankydan commented on libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/TransactionMetadata.kt:
```diff
@@ -36,11 +36,11 @@ class TransactionMetadata(private val properties: LinkedHashMap<String, Any>) {
-        try {
-            return Integer.parseInt(version)
+        return try {
+            version.toInt()
```
Theres a similar change to make further down in the file

omfR3 replied:
Thanks, missed that.

omfR3 committed:
CORE-7147: (Transaction metadata) Follow through on an earlier refactoring

mcgovc reviewed (approved):
LGTM - build changes
